### ISSUE TITLE
How to port/execute a test(s) to a generic machine that can build the model, port the UFS WM on AWS cloud

### ISSUE THREAD

**Issue by aerorahul** [ufs-community/ufs-weather-model#263]
## Description
The build of the UFS uses cmake, which makes building of the UFS model on a generic machine a breeze.
To execute the model several files (input data, model fix files, etc.) need to be staged.

At a minimum, a low resolution forecast run of the model in weather (C96), S2S (1 deg ocean) and data-atmosphere mode would be helpful for developers who don't have access to machines as well as facilitate community development.

## Solution
Documentation with clear list of files that need to be obtained and staged.  Where to get those files.
Documentation for scripts (`rt.sh`, `utest`, `compile.sh`, etc.) that need to be edited to point to the local copied data.

**Comment by climbfuji** [comment]
I have been given free storage on Amazon S3 for such purposes (and also for staging regression test input data for running the regression tests in the cloud). I can make the required data available at no cost to the users or NOAA.

**Comment by junwang-noaa** [comment]
Awesome, Dom! Please let us know when it is available and send us
instructions on how general developers can get the data.

**Comment by climbfuji** [comment]
I just need to know what to upload and where to grab it from. I have the bucket ready, it will be a simply https://... link to wget/curl or copy&paste into a browser.

**Comment by junwang-noaa** [comment]
@climbfuji do you have the low resolution input data for standalone fv3, s2s and datm RT on s3? Can you provide a link here? Thanks

**Comment by arunchawla-NOAA** [comment]
@junwang-noaa is there work going on for this issue ? Can we consider this closed ?

**Comment by junwang-noaa** [comment]
Yes, I think we can close the issue.

**Comment by aerorahul** [comment]
@junwang-noaa 
Do we have instructions to refer to where one can find the test data and run a test?

**Comment by junwang-noaa** [comment]
@climbfuji Do you have any documentation on how to access data on s3?

**Comment by climbfuji** [comment]
No. This has not been maintained for a while now, because it never got used despite me keeping the data in sync.

**Comment by DeniseWorthen** [comment]
It seems S3 is now being maintained by EPIC (see Issue #1569)? Can we close this?

**Comment by arunchawla-NOAA** [comment]
@jkbk2004 bringing to your attention to address or close

**Comment by jkbk2004** [comment]
Yes, files must be available https://noaa-ufs-regtests-pds.s3.amazonaws.com/. But user guide update is the thing to follow on, right? EPIC can follow on with @gspetro-NOAA

**Comment by sadeghitabas** [comment]
I've prepared a brief guide on how to:
1- Install UFS WM and its dependencies using spack-stack on NOAA AWS machine.
2- Get input data and baselines from Amazon S3 bucket and put them in the right directory to be used by UFS WM.
3- Update UFS WM modulefiles and other scripts to be able to run a regression test on NOAA AWS cloud machine. 

Please see the [google doc](https://docs.google.com/document/d/1K6n4yaJz5ylRtD4XEnyKALSilbdFTUbbEPvDJmcDSos/edit?usp=sharing) for more information.

**Comment by JessicaMeixner-NOAA** [comment]
Is it possible to copy from the Amazon S3 bucket the input data to anywhere else?

**Comment by jkbk2004** [comment]
Yes,` wget` can be used.

**Comment by jkbk2004** [comment]
`wget http://noaa-ufs-regtests-pds.s3.amazonaws.com/develop-20230126/INTEL/atmaero_control_p8/sfcf024.nc`

**Comment by junwang-noaa** [comment]
@SadeghTabas-NOAA Thanks for the nice work! @aerorahul FYI.

**Comment by jkbk2004** [comment]
@gspetro-NOAA we can coordinate to update user guide to add a subsection for PW-Cloud option. can we tag-up with @sadegh @zach1221

**Comment by jkbk2004** [comment]
> @gspetro-NOAA we can coordinate to update user guide to add a subsection for PW-Cloud option. can we tag-up with @sadegh @zach1221 ?

**Comment by zach1221** [comment]
I have put a rough draft document together for how to build/run UFS-WM on Parallel Works Azure, AWS, GCP. That may be helpful for the subsection.

**Comment by gspetro-NOAA** [comment]
@zach1221 If you send me the doc, I can start inputting the info into the WM docs rst files. I'm starting an update that should wrap up before the end of the PI. If you, @jkbk2004 , or @SadeghTabas-NOAA feel further discussion is required/desirable, feel free to add a meeting on my calendar.

**Comment by zach1221** [comment]
@gspetro-NOAA I just gave you access to the doc in the CM share drive. Feel free to add anything or reach out if you have any questions.

**Comment by sadeghitabas** [comment]
@gspetro-NOAA Please let me know if you need any help with that. I'd be happy to help. Thanks,

**Comment by junwang-noaa** [comment]
@zach1221 @gspetro-NOAA Do you have the document ready for us to review? Are they different from the doc Sadegh added in this issue? How long will the document can be committed? If your document is the same as Sadegh and can get it committed within next two weeks, we can leave this to you. Otherwise we plan to commit the document Sadeh has on NOAA AWS cloud because lots of developers are waiting for it, you can update the document with another PR to include all three cloud support, is that working for you? Thanks

**Comment by zach1221** [comment]
@junwang-noaa I think the basic scope of the Sadegh's document and mine is the same, however his document mentions spack-stack and mine hpc-stack.  I can have a draft PR for all three cloud environments, with Sadegh's work combined in, ready by end of next week.

**Comment by jkbk2004** [comment]
@ulmononian FYI

**Comment by gspetro-NOAA** [comment]
@SadeghTabas-NOAA In your google doc, where it says to issue the "module unuse" statements, are those only necessary on systems where Cray is default? And is there a way to provide the information users need without hardcoded paths so that it's applicable to any Linux system? For example, how would they determine if there's a set of modules loaded that require a `module unuse`?

**Comment by ulmononian** [comment]
> @SadeghTabas-NOAA In your google doc, where it says to issue the "module unuse" statements, are those only necessary on systems where Cray is default? And is there a way to provide the information users need without hardcoded paths so that it's applicable to any Linux system? For example, how would they determine if there's a set of modules loaded that require a `module unuse`? 

users should refer to the spack stack docs for the prerequisite load / unload commands for each site.

https://spack-stack.readthedocs.io/en/latest/Platforms.html#pre-configured-sites

please note that the sequence of load/unload commands differs if users are (i) building a new spack stack env or (ii) just loading an existing env for use.

**Comment by sadeghitabas** [comment]
> @SadeghTabas-NOAA In your google doc, where it says to issue the "module unuse" statements, are those only necessary on systems where Cray is default? And is there a way to provide the information users need without hardcoded paths so that it's applicable to any Linux system? For example, how would they determine if there's a set of modules loaded that require a `module unuse`?

@ulmononian Thank you for your answer. 
@gspetro-NOAA As @ulmononian suggested, this is just for building spack-stack on cloud (AWS, Azure and GCP) and it needs to be done for all three cloud platforms based on spack-stack documentation. For other machines please refer to the spack-stack documentation link provided by @ulmononian.
Thanks

**Comment by gspetro-NOAA** [comment]
@junwang-noaa To clarify, are developers waiting on the whole of @SadeghTabas-NOAA 's doc, or just specific sections, like where to find the data?

**Comment by junwang-noaa** [comment]
We hope to have the step by step instructions on running a UFS RT test on
cloud documented in ufs wm doc. Sadegh already has the document, do we need
anything else to finish the documentation?

**Comment by gspetro-NOAA** [comment]
@junwang-noaa I have been working on documentation in the .rst files, but I'm trying to understand the purpose/scope of this issue, since @SadeghTabas-NOAA and @zach1221 's docs both start with spack-stack/HPC-Stack installation. Those sections seem like they shouldn't be included, since there is separate documentation for spack-stack/HPC-Stack that users can/should refer to. It might be best to start in the WM docs from the assumption that users have already downloaded/installed/loaded prerequisite software. Otherwise, we're basically including documentation for another repo in our repo.

**Comment by junwang-noaa** [comment]
If we can provide links of the prerequisite software instructions, that would be good. Then we just need to focus on UFS WM. I will let @SadeghTabas-NOAA to confirm too.

**Comment by sadeghitabas** [comment]
@junwang-noaa @gspetro-NOAA There is instruction for building spack-stack and we can refer to that one. However, I had some difficulties for the first time building it. Also, as @ulmononian mentioned in [PR1667](https://github.com/ufs-community/ufs-weather-model/pull/1667) they are installing spack-stack beta unified environment (UE) on official tier-1 machines and the google doc I prepared may be out dated as the spack-stack v1.3 UE comes in. So only referring to the spack-stack instructions would be enough.

**Comment by sadeghitabas** [comment]
@gspetro-NOAA I've updated UFS-WM cloud module files format to be similar to other machines. I've updated the google doc accordingly.

**Comment by ulmononian** [comment]
> @junwang-noaa @gspetro-NOAA There is instruction for building spack-stack and we can refer to that one. However, I had some difficulties for the first time building it. Also, as @ulmononian mentioned in [PR1667](https://github.com/ufs-community/ufs-weather-model/pull/1667) they are installing spack-stack beta unified environment (UE) on official tier-1 machines and the google doc I prepared may be out dated as the spack-stack v1.3 UE comes in. So only referring to the spack-stack instructions would be enough.

until the request @SadeghTabas-NOAA made to the parallel works group regarding the creation of a shared `/contrib` directory for users of different projects (e.g. {ca,cg,cz}-epic & ca-ufs-cpldcld), and we have to maintain two stacks that are "blind" to each other, i would recommend that the unified-environment be installed in each project's contrib space so that any testing of the weather model RTs is done against a consistent stack. although UE installations on tier-1 machines takes precedence w/ the spack-stack/1.3.0 release, @mark-a-potts is working on official installations of the UE on epic project `/contrib` spaces. my suggestion is that this is also done on the ca-ufs-cpldcld project `/contrib` space when spack-stack/1.3.0 is finalized.

**Comment by sadeghitabas** [comment]
This is a [video on how to port ufswm for a generic machine (pw-aws)](https://drive.google.com/file/d/1P_cHST79RKsa3BsbqvyHd2NCQtP6dRCd/view?usp=share_link) recorded on the EIB meeting. I also provided a [google doc](https://docs.google.com/document/d/1K6n4yaJz5ylRtD4XEnyKALSilbdFTUbbEPvDJmcDSos/edit?usp=sharing) for that. In addition, PR[#1667](https://github.com/ufs-community/ufs-weather-model/pull/1667) corresponding to ufswm compatibility of cloud machines (aws, gcp and azure) is committed. I think now we can close this issue. Thanks.
